### PR TITLE
8300256: C2: vectorization is sometimes skipped on loops where it would succeed

### PR DESCRIPTION
**Pull request by rwestrel**
Vectorization for a counted loop cl only proceeds if
cl->range_checks_present() returns true. The result of that method is
computed lazily and its result cached in the CountedLoopNode and never
re-computed. If PhaseIdealLoop::do_range_check() returns 0 then the
result of that computation is overwritten (no range checks
present). PhaseIdealLoop::do_range_check() counts the number of tests
present in the loop body (which is really what range_checks_present()
is about) and decrements that count for every check it eliminates
except if it's not a comparison with a LoadRange (for a reason that I
don't understand). In the case of the test (a pattern from a
ByteBuffer benchmark), not all tests are with a LoadRange. As a
result, PhaseIdealLoop::do_range_check() returns non zero even though
it eliminates all tests. As a result, vectorization is never
attempted.

There doesn't seem to be a value in caching the result of
range_checks_present() in CountedLoopNode. It's not that expensive to
compute, it's only used during loop opts and it's really hard to keep
in sync with whether the loop has still tests: several different
transformations could remove a test. What I propose instead is to keep
roughly the same approach (compute the result lazily and cache it so
it doesn't have to be re-computed) but to store it on the
IdealLoopTree instead (so it's recomputed on every loop opts pass and
there's no risk that it becomes out of sync).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300256](https://bugs.openjdk.org/browse/JDK-8300256): C2: vectorization is sometimes skipped on loops where it would succeed

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [2140b251](https://git.openjdk.org/jdk/pull/12116/files/2140b251d201a427b31a23d4d456e9fcb5868cf9)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [c48ccd47](https://git.openjdk.org/jdk/pull/12116/files/c48ccd47c2d8f79df0aee9fcca8ae7b5243004c5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12116/head:pull/12116` \
`$ git checkout pull/12116`

Update a local copy of the PR: \
`$ git checkout pull/12116` \
`$ git pull https://git.openjdk.org/jdk pull/12116/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12116`

View PR using the GUI difftool: \
`$ git pr show -t 12116`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12116.diff">https://git.openjdk.org/jdk/pull/12116.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -3990,7 +3990,6 @@ void IdealLoopTree::dump_head() const {
     if (cl->is_post_loop()) tty->print(" post");
     if (cl->is_reduction_loop()) tty->print(" reduction");
     if (cl->is_vectorized_loop()) tty->print(" vector");
-    if (cl->range_checks_present()) tty->print(" rc ");
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:3993]: Can you use new `range_checks_present()` here? The information is still useful.

**Review by vnkozlov — COMMENTED** [review]
Just one comment.

File: src/hotspot/share/opto/loopnode.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopnode.cpp:3993]: Thanks for reviewing this.
Indeed, you're right. I updated the change.

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Maybe add a `@bug 8300256` to the test.

**Comment by rwestrel** [comment]
> Looks good to me. Maybe add a `@bug 8300256` to the test.

Thanks for the review. Yes, I will do that.

**Review by TobiHartmann — APPROVED** [review]

**Comment by rwestrel** [comment]
FTR, the test fails on x86 (32 bits) for some reason. I changed the test so it doesn't run on x86 (32 bits).

**Pull request closed without merging**